storage/storage_image.go

```diff
@@ -0,0 +1,554 @@
+			}
+			return errorBlobInfo, ErrBlobDigestMismatch
+		}
+		// If we didn't get a name, we might as well assign one using the hash.
+		if digest == "" {
+			digest = "sha256:" + hex.EncodeToString(hash[:])
+		}
+		// Record that this blob is a layer.
+		s.Layers[digest] = id
+		s.BlobList = append(s.BlobList, digest)
+		if layer != nil {
+			logrus.Debugf("blob %q imported as a filesystem layer %q", blobinfo.Digest, digest)
+		} else {
+			logrus.Debugf("layer blob %q already present", blobinfo.Digest)
+		}
+	} else {
+		// It's just data.  Finish scanning it in, check that our
+		// computed sha256sum matches the digest, and store it, but
+		// leave it out of the blob-to-layer-ID map so that we can tell
+		// that it's not a layer.
+		blob, err := ioutil.ReadAll(multi)
+		if err != nil && err != io.EOF {
+			return errorBlobInfo, err
+		}
+		actualSize = int64(len(blob))
+		hash = hasher.Sum(nil)
+		// If the ID was a digest, verify that our computed sha256sum
+		// matches it. */
+		if strings.HasPrefix(blobinfo.Digest, "sha256:") && blobinfo.Digest != "sha256:"+hex.EncodeToString(hash[:]) {
+			logrus.Debugf("blob %q digests to %q, rejecting", blobinfo.Digest, hex.EncodeToString(hash[:]))
+			return errorBlobInfo, ErrBlobDigestMismatch
+		}
+		// If we didn't get a name, we might as well assign one using the hash.
+		if digest == "" {
+			digest = "sha256:" + hex.EncodeToString(hash[:])
+		}
+		// Save the blob for when we Commit().
+		s.BlobData[digest] = blob
+		s.BlobList = append(s.BlobList, digest)
+		logrus.Debugf("blob %q imported as opaque data %q", blobinfo.Digest, digest)
+	}
+	return types.BlobInfo{
+		Digest: digest,
+		Size:   actualSize,
+	}, nil
+}
+
+func (s *storageImage) Commit() error {
+	if s.ID != "" {
+		// We started with an image ID, or we've already registered
+		// this one and gotten one, so no need to do anything more.
+		if img, err := s.store.GetImage(s.ID); img != nil && err == nil {
+			return nil
+		}
+	}
+	lastLayer := ""
+	if len(s.BlobList) > 0 {
+		for _, blob := range s.BlobList {
+			if layerID, ok := s.Layers[blob]; ok {
+				lastLayer = layerID
+			}
+		}
+	}
+	img, err := s.store.CreateImage(s.ID, nil, lastLayer, "")
+	if err != nil {
+		return err
+	}
+	logrus.Debugf("created new image ID %q", img.ID)
+	if s.Tag != "" {
+		// We started with an image name rather than an ID, so move the
+		// name to this image.
+		if err := s.store.SetNames(img.ID, []string{s.Tag}); err != nil {
+			return err
+		}
+		logrus.Debugf("set name of image %q to %q", img.ID, s.Tag)
+	}
+	// Save the blob data to disk, and drop the contents from memory.
+	keys := []string{}
+	for k, v := range s.BlobData {
+		if err := s.store.SetImageBigData(img.ID, k, v); err != nil {
+			return err
+		}
+		keys = append(keys, k)
+	}
+	for _, key := range keys {
+		delete(s.BlobData, key)
+	}
+	s.ID = img.ID
+	return nil
+}
+
+func (s *storageImage) PutManifest(manifest []byte) error {
+	if err := s.Commit(); err != nil {
```

Comment:
Changing the intermediate steps to avoid saving partial progress, except for layer blobs, which we probably don't want to buffer in memory, so that Commit() will actually be doing more of the work, and to not have the intermediate functions call Commit().